_posts/2020-10-07-torch-modules/torch_modules.Rmd

```diff
@@ -0,0 +1,429 @@
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchmodules
+date: 10-07-2020
+categories:
+  - Torch
+  - R
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+editor_options: 
+  markdown: 
+    wrap: 72
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+[Last
+week](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch),
+we started learning about `torch` basics by coding a simple neural
+network from scratch, making use of just a single of `torch`'s features:
+*tensors*. [On
+Monday](https://blogs.rstudio.com/ai/posts/2020-10-05-torch-network-with-autograd),
+we immensely simplified the task, replacing manual backpropagation with
+*autograd*. Today, we *modularize* the network - in both the habitual
+and a very literal sense: Low-level matrix operations are swapped out
+for `torch` `module`s.
+
+## Modules
+
+From other frameworks (Keras, say), you may be used to distinguishing
+between *models* and *layers*. In `torch`, both are instances of
+`nn_Module()`, and thus, have some methods in common. For those thinking
+in terms of "models" and "layers", I'm artificially splitting up this
+section into two parts, for clarity of exposition. In reality though,
+there is no dichotomy: New modules may be composed of existing ones up
+to arbitrary levels of recursion.
+
+### Base modules ("layers")
+
+Instead of writing out an affine operation by hand -- `x$mm(w1) + b1`,
+say --, as we've been doing so far, we can create a linear module. The
+following snippet instantiates a linear layer that expects three-feature
+inputs and returns a single output per observation:
+
+```{r}
+library(torch)
+l <- nn_linear(3, 1)
+```
+
+The module has two parameters, "weight" and "bias". Both have already
+been initialized:
+
+```{r}
+l$parameters
+```
+
+    $weight
+    torch_tensor 
+    -0.0385  0.1412 -0.5436
+    [ CPUFloatType{1,3} ]
+
+    $bias
+    torch_tensor 
+    -0.1950
+    [ CPUFloatType{1} ]
+
+Modules are callable; calling a module executes its `forward()` method,
+which for a linear layer just matrix-multiplies input and weights and
+adds the bias.
+
+Let's try this:
+
+```{r}
+data  <- torch_randn(10, 3)
+out <- l(data)
+```
+
+Unsurprisingly, `out` now holds some data:
+
+```{r}
+out$data()
+```
+
+    torch_tensor 
+     0.2711
+    -1.8151
+    -0.0073
+     0.1876
+    -0.0930
+     0.7498
+    -0.2332
+    -0.0428
+     0.3849
+    -0.2618
+    [ CPUFloatType{10,1} ]
+
+In addition though, it knows what will need to be done, should ever it
+be asked to calculate gradients:
+
+```{r}
+out$grad_fn
+```
+
+    AddmmBackward
+
+(You may remember from the preceding post that when we work with tensors
+directly, we need to create them passing `requires_grad = TRUE` to
+achieve this. With modules, `torch` correctly assumes that we'll want to
+perform backpropagation at some point.)
```

Comment:
```suggestion
perform backpropagation at some point.
```

Reply:
Having a stand-alone parenthetical seems off to me. You could put it as an aside, if you think it's out of place inline with everything else.

Reply:
oh yeah, you're spotting everything :-)

Thinking about it, I even changed this to a paragraph starting "Note the difference..." :-)